Amelioration for networked multiplayer on single machine

j-lag opened the pull request:
- Allow joystick to be fetch if windows has no focus in SDL input mode
- Add video settings:
      `-window-pos=<x>,<y>`            Position [Default: centered]
      `-borderless`             Windowed mode with no border
These 2 settings are usefull when setting up a multiplayer game on one machine.

Example for 4 windows in fullHD:
`start /D"Master" Supermodel.exe -input-system=sdl dayto2pe.zip -res=960,540 -borderless -window-pos=0,0`
`start /D"Slave1" Supermodel.exe -input-system=sdl dayto2pe.zip -res=960,540 -borderless -window-pos=960,0`
`start /D"Slave2" Supermodel.exe -input-system=sdl dayto2pe.zip -res=960,540 -borderless -window-pos=0,540`
`start /D"Slave3" Supermodel.exe -input-system=sdl dayto2pe.zip -res=960,540 -borderless -window-pos=960,540`

trzy commented on Src/OSD/SDL/Main.cpp:
```diff
@@ -1086,8 +1086,15 @@ int Supermodel(const Game &game, ROMSet *rom_set, IEmulator *Model3, CInputs *In
   SDL_SetWindowSize(s_window, totalXRes, totalYRes);
-  SDL_SetWindowPosition(s_window, SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED);
+
+  if ( !s_runtime_config["Xpos"].Empty() && !s_runtime_config["Xpos"].Empty())
```
I think you mean to check Xpos and Ypos. I have a few requests here:

1. Remove the leading whitespace in the leading parenthesis.
2. Add braces to the if and else statements. I know this isn't how the rest of the code base looks (yet) but I'd like us to move to that style as it is considered best practice in all C-like language these days to avoid unbraced if-statements and single-line loops.
3. Please change Xpos and Ypos to "WindowXPosition" and "WindowYPosition" to match the verbose format of most other config options.

trzy commented on Src/OSD/SDL/Main.cpp:
```diff
@@ -1086,8 +1086,15 @@ int Supermodel(const Game &game, ROMSet *rom_set, IEmulator *Model3, CInputs *In
+  else  
+    SDL_SetWindowPosition(s_window, SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED);
   
+  if (s_runtime_config["BorderLess"].ValueAs<bool>())
```
1. Braces around the statement.
2. BorderLess -> BorderlessWindow

trzy commented on Src/OSD/SDL/Main.cpp:
```diff
@@ -1708,7 +1719,9 @@ static void Help(void)
   puts("");
   puts("Video Options:");
   puts("  -res=<x>,<y>            Resolution [Default: 496,384]");
+  puts("  -pos=<x>,<y>            Position [Default: centered]");  
```
Change -pos to -window-pos please.

trzy commented on Src/OSD/SDL/Main.cpp:
```diff
@@ -1936,6 +1950,31 @@ static ParsedCommandLine ParseCommandLine(int argc, char **argv)
           }
         }
       }
+      else if (arg == "-pos" || arg.find("-pos=") == 0)
```
When changing -pos -> -window-pos don't forget to update all occurrences here.

trzy commented on Src/OSD/SDL/Main.cpp:
```diff
@@ -1936,6 +1950,31 @@ static ParsedCommandLine ParseCommandLine(int argc, char **argv)
+          std::vector<std::string> parts = Util::Format(arg).Split('=');
+          if (parts.size() != 2)
+          {
+              ErrorLog("'-pos' requires both a X and Y (e.g., '-pos=10,0').");
```
Nitpick: "both a X and Y" -> "both an X and Y position"

trzy commented on Src/OSD/SDL/Main.cpp:
```diff
@@ -1936,6 +1950,31 @@ static ParsedCommandLine ParseCommandLine(int argc, char **argv)
+              }
+              else
+              {
+                  ErrorLog("'-pos' requires both a X and Y (e.g., '-pos=10,0').");
```
Same here.

j-lag commented:
No worries!
Changes done in comit 722c70bf865892fa64cc369e368e2fe170b1b403 !
best regards.

Le jeu. 2 mars 2023 à 20:36, trzy ***@***.***> a écrit :

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In Src/OSD/SDL/Main.cpp
> <https://github.com/trzy/Supermodel/pull/64#discussion_r1123610667>:
>
> > @@ -1086,8 +1086,15 @@ int Supermodel(const Game &game, ROMSet *rom_set, IEmulator *Model3, CInputs *In
>    sprintf(baseTitleStr, "Supermodel - %s", game.title.c_str());
>    SDL_SetWindowTitle(s_window, baseTitleStr);
>    SDL_SetWindowSize(s_window, totalXRes, totalYRes);
> -  SDL_SetWindowPosition(s_window, SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED);
> +
> +  if ( !s_runtime_config["Xpos"].Empty() && !s_runtime_config["Xpos"].Empty())
>
> I think you mean to check Xpos and Ypos. I have a few requests here:
>
>    1. Remove the leading whitespace in the leading parenthesis.
>    2. Add braces to the if and else statements. I know this isn't how the
>    rest of the code base looks (yet) but I'd like us to move to that style as
>    it is considered best practice in all C-like language these days to avoid
>    unbraced if-statements and single-line loops.
>    3. Please change Xpos and Ypos to "WindowXPosition" and
>    "WindowYPosition" to match the verbose format of most other config options.
>
> ------------------------------
>
> In Src/OSD/SDL/Main.cpp
> <https://github.com/trzy/Supermodel/pull/64#discussion_r1123611228>:
>
> >
> +  if (s_runtime_config["BorderLess"].ValueAs<bool>())
>
>
>    1. Braces around the statement.
>    2. BorderLess -> BorderlessWindow
>
> ------------------------------
>
> In Src/OSD/SDL/Main.cpp
> <https://github.com/trzy/Supermodel/pull/64#discussion_r1123611512>:
>
> > @@ -1708,7 +1719,9 @@ static void Help(void)
>    puts("");
>    puts("Video Options:");
>    puts("  -res=<x>,<y>            Resolution [Default: 496,384]");
> +  puts("  -pos=<x>,<y>            Position [Default: centered]");
>
> Change -pos to -window-pos please.
> ------------------------------
>
> In Src/OSD/SDL/Main.cpp
> <https://github.com/trzy/Supermodel/pull/64#discussion_r1123613128>:
>
> > @@ -1936,6 +1950,31 @@ static ParsedCommandLine ParseCommandLine(int argc, char **argv)
>            }
>          }
>        }
> +      else if (arg == "-pos" || arg.find("-pos=") == 0)
>
> When changing -pos -> -window-pos don't forget to update all occurrences
> here.
> ------------------------------
>
> In Src/OSD/SDL/Main.cpp
> <https://github.com/trzy/Supermodel/pull/64#discussion_r1123613640>:
>
> > @@ -1936,6 +1950,31 @@ static ParsedCommandLine ParseCommandLine(int argc, char **argv)
>            }
>          }
>        }
> +      else if (arg == "-pos" || arg.find("-pos=") == 0)
> +      {
> +          std::vector<std::string> parts = Util::Format(arg).Split('=');
> +          if (parts.size() != 2)
> +          {
> +              ErrorLog("'-pos' requires both a X and Y (e.g., '-pos=10,0').");
>
> Nitpick: "both a X and Y" -> "both an X and Y position"
> ------------------------------
>
> In Src/OSD/SDL/Main.cpp
> <https://github.com/trzy/Supermodel/pull/64#discussion_r1123613774>:
>
> > +              ErrorLog("'-pos' requires both a X and Y (e.g., '-pos=10,0').");
> +              cmd_line.error = true;
> +          }
> +          else
> +          {
> +              unsigned  x, y;
> +              if (2 == sscanf(&argv[i][4], "=%u,%u", &x, &y))
> +              {
> +                  std::string xres = Util::Format() << x;
> +                  std::string yres = Util::Format() << y;
> +                  cmd_line.config.Set("Xpos", xres);
> +                  cmd_line.config.Set("Ypos", yres);
> +              }
> +              else
> +              {
> +                  ErrorLog("'-pos' requires both a X and Y (e.g., '-pos=10,0').");
>
> Same here.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/trzy/Supermodel/pull/64#pullrequestreview-1322489484>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC5ITU66S7OV4KHRAFP6K5TW2DZDNANCNFSM6AAAAAAVNLRZDA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

trzy commented on Src/OSD/SDL/Main.cpp:
```diff
@@ -1708,7 +1725,9 @@ static void Help(void)
   puts("");
   puts("Video Options:");
   puts("  -res=<x>,<y>            Resolution [Default: 496,384]");
+  puts("  -window-pos=<x>,<y>     Position [Default: centered]");  
```
Change description to: Window position [Default: centered]

trzy commented on Src/OSD/SDL/Main.cpp:
```diff
@@ -1086,8 +1086,21 @@ int Supermodel(const Game &game, ROMSet *rom_set, IEmulator *Model3, CInputs *In
+  }
+  else
+  {
+      SDL_SetWindowPosition(s_window, SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED);
```
Looks like an indentation issue on either this line and/or the if-statement above.

trzy reviewed (changes requested):
Looks great and thank you for the changes! Just two very small changes (an indentation issue and adding a word the help description) and we should be good to merge.

j-lag commented:
Done!
Yea my visual studio and I ar used to 4 spaces indents when your code use 2
spaces indents, thats why :)

Le sam. 4 mars 2023 à 01:09, trzy ***@***.***> a écrit :

> ***@***.**** requested changes on this pull request.
>
> Looks great and thank you for the changes! Just two very small changes (an
> indentation issue and adding a word the help description) and we should be
> good to merge.
> ------------------------------
>
> In Src/OSD/SDL/Main.cpp
> <https://github.com/trzy/Supermodel/pull/64#discussion_r1125154375>:
>
> > @@ -1708,7 +1725,9 @@ static void Help(void)
>    puts("");
>    puts("Video Options:");
>    puts("  -res=<x>,<y>            Resolution [Default: 496,384]");
> +  puts("  -window-pos=<x>,<y>     Position [Default: centered]");
>
> Change description to: Window position [Default: centered]
> ------------------------------
>
> In Src/OSD/SDL/Main.cpp
> <https://github.com/trzy/Supermodel/pull/64#discussion_r1125154678>:
>
> > @@ -1086,8 +1086,21 @@ int Supermodel(const Game &game, ROMSet *rom_set, IEmulator *Model3, CInputs *In
>    sprintf(baseTitleStr, "Supermodel - %s", game.title.c_str());
>    SDL_SetWindowTitle(s_window, baseTitleStr);
>    SDL_SetWindowSize(s_window, totalXRes, totalYRes);
> -  SDL_SetWindowPosition(s_window, SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED);
> +
> +  if (!s_runtime_config["WindowXPosition"].Empty() && !s_runtime_config["WindowYPosition"].Empty())
> +  {
> +    SDL_SetWindowPosition(s_window, s_runtime_config["WindowXPosition"].ValueAs<unsigned>(), s_runtime_config["WindowYPosition"].ValueAs<unsigned>());
> +  }
> +  else
> +  {
> +      SDL_SetWindowPosition(s_window, SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED);
>
> Looks like an indentation issue on either this line and/or the
> if-statement above.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/trzy/Supermodel/pull/64#pullrequestreview-1324776924>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC5ITUZKOGCG7PQE5XM6XPTW2KB3DANCNFSM6AAAAAAVNLRZDA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

trzy commented:
Quick question: you set the default values to NA but that means the line that checks for whether those are empty will always return false. I didn’t anticipate this situation where there is a purely optional config value. Let me test this in a few hours and get back to you. The config system doesn’t necessary need to have defaults set. 

Supermodel will try to cast “NA” to unsigned which will probably be 0. We don’t have a convention of using different types for the same config key. There are really three ways to handle this: 1) a special magic number is used (like SDL has for the window centered positions), which isn’t great, or I think 2) we can avoid setting these values in the default initialization. The values could 3) also be made signed ints and negative values can be interpreted to center the window.

Which do you prefer? I am leaning toward option 2, with options 3 and 1 as my next choices.

If we go with option 2 there should be a comment added near the empty check that default values for WindowXPos and WindowYPos do not exist so empty must be checked for explicitly.

trzy commented:
Actually I thought of an even better way: the config system should support “empty” values. Empty() means the config key doesn’t exist. We should have a method that checks if it is Empty() or if the value is an Empty() string (0 length or containing only white space). EmptyOrBlank().

I’ll paste some code in an hour or so to show exactly what I mean.

I can then do a follow on PR that changes Empty() too !Exists() and EmptyOrBlank() to Empty(). I think that is probably a more clear design?

j-lag commented:
Hi!
Actually I first wrote something along the line of if(strcmp(posx , "NA")
!=0 ...)
But when I tested it it crash when not set.
I have not look further into your default param system but I found that the
parameter is empty when not set by the cmdline. Bottom line : it work as
is...

Le sam. 4 mars 2023, 18:53, trzy ***@***.***> a écrit :

> Quick question: you set the default values to NA but that means the line
> that checks for whether those are empty will always return false. I didn’t
> anticipate this situation where there is a purely optional config value.
> Let me test this in a few hours and get back to you. The config system
> doesn’t necessary need to have defaults set.
>
> Supermodel will try to cast “NA” to unsigned which will probably be 0. We
> don’t have a convention of using different types for the same config key.
> There are really three ways to handle this: 1) a special magic number is
> used (like SDL has for the window centered positions), which isn’t great,
> or I think 2) we can avoid setting these values in the default
> initialization. The values could 3) also be made signed ints and negative
> values can be interpreted to center the window.
>
> Which do you prefer? I am leaning toward option 2, with options 3 and 1 as
> my next choices.
>
> If we go with option 2 there should be a comment added near the empty
> check that default values for WindowXPos and WindowYPos do not exist so
> empty must be checked for explicitly.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/trzy/Supermodel/pull/64#issuecomment-1454821164>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC5ITUZYRV42JENJN53XAODW2N6TLANCNFSM6AAAAAAVNLRZDA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

j-lag commented:
Ok fine, I will not have a lot of time to test in the upcoming days but I
will happily integrate any change you want.

Le sam. 4 mars 2023, 19:04, trzy ***@***.***> a écrit :

> Actually I thought of an even better way: the config system should support
> “empty” values. Empty() means the config key doesn’t exist. We should have
> a method that checks if it is Empty() or if the value is an Empty() string
> (0 length or containing only white space). EmptyOrBlank().
>
> I’ll paste some code in an hour or so to show exactly what I mean.
>
> I can then do a follow on PR that changes Empty() too !Exists() and
> EmptyOrBlank() to Empty(). I think that is probably a more clear design?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/trzy/Supermodel/pull/64#issuecomment-1454824923>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC5ITU2EC3CGKUHBNDHZ2TTW2N72ZANCNFSM6AAAAAAVNLRZDA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

trzy commented:
Ok, I have made some fixes. Please see my branch (bart/windowpos) and cherry pick the two commits I made into yours (or rebase yours on mine):

```
* bb8b31e  (HEAD -> bart/windowpos, origin/bart/windowpos) (2023-03-04 11:59:49) Bart Trzynadlowski Fixed window position config storage and command line parsing
* 43833a2  (2023-03-04 11:58:52) Bart Trzynadlowski Config nodes: added the ability to clear out node values or create empty leaf nodes
```
Three files were changed: Main.cpp, NewConfig.cpp, NewConfig.h. Here's what I did:

- Made the values ints instead of unsigned because that's what SDL takes.
- Check each of WindowXPosition, WindowYPosition individually and if it does not exist, use SDL_WINDOWPOS_CENTERED for only that axis. This means if you specify one of these options in the config file but not the other, it will still work, but only one axis will be set and the other will be centered.
- Updated NewConfig.cpp/NewConfig.h to support creating empty nodes. The key will be present in the config tree but there will be no value, so Empty() returns true and Exists() returns false. No change to the API as a result of this!

This is different than what I proposed. I had to refresh myself on my complicated config system :) But it's a better solution. Basically, the config system works like this:

- It's a nested tree of keys and values.
- Values can be almost any type. They are *not* strings underneath (I had forgotten about this). In fact, they are whatever type is used to initially set them.
- When loading from INI or XML, the values will indeed be loaded as strings. But otherwise, they can be set directly to ints or floats or whatever, and that will be the underlying representation. ValueAs<T>() only performs a conversion if it is not already stored as T.
- ValueAs<T>() will throw an exception if there is no value in the node.
- ValueAsDefault<T>(T default) will not, instead it will behave like ValueAs<T>() if a value exists, otherwise it will return the default you specify. It never throws an exception.
- I've added a SetEmpty() method that can be used to create empty keys or clear existing ones and I use this for WindowXPosition, etc.

Hopefully this makes more sense.

j-lag commented:
Cool I will check that tomorrow after work!


Le sam. 4 mars 2023, 21:07, trzy ***@***.***> a écrit :

> Ok, I have made some fixes. Please see my branch (bart/windowpos) and
> cherry pick the two commits I made into yours (or rebase yours on mine):
>
> * bb8b31e  (HEAD -> bart/windowpos, origin/bart/windowpos) (2023-03-04 11:59:49) Bart Trzynadlowski Fixed window position config storage and command line parsing
> * 43833a2  (2023-03-04 11:58:52) Bart Trzynadlowski Config nodes: added the ability to clear out node values or create empty leaf nodes
>
> Three files were changed: Main.cpp, NewConfig.cpp, NewConfig.h. Here's
> what I did:
>
>    - Made the values ints instead of unsigned because that's what SDL
>    takes.
>    - Check each of WindowXPosition, WindowYPosition individually and if
>    it does not exist, use SDL_WINDOWPOS_CENTERED for only that axis. This
>    means if you specify one of these options in the config file but not the
>    other, it will still work, but only one axis will be set and the other will
>    be centered.
>    - Updated NewConfig.cpp/NewConfig.h to support creating empty nodes.
>    The key will be present in the config tree but there will be no value, so
>    Empty() returns true and Exists() returns false. No change to the API as a
>    result of this!
>
> This is different than what I proposed. I had to refresh myself on my
> complicated config system :) But it's a better solution. Basically, the
> config system works like this:
>
>    - It's a nested tree of keys and values.
>    - Values can be almost any type. They are *not* strings underneath (I
>    had forgotten about this). In fact, they are whatever type is used to
>    initially set them.
>    - When loading from INI or XML, the values will indeed be loaded as
>    strings. But otherwise, they can be set directly to ints or floats or
>    whatever, and that will be the underlying representation. ValueAs() only
>    performs a conversion if it is not already stored as T.
>    - ValueAs() will throw an exception if there is no value in the node.
>    - ValueAsDefault(T default) will not, instead it will behave like
>    ValueAs() if a value exists, otherwise it will return the default you
>    specify. It never throws an exception.
>    - I've added a SetEmpty() method that can be used to create empty keys
>    or clear existing ones and I use this for WindowXPosition, etc.
>
> Hopefully this makes more sense.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/trzy/Supermodel/pull/64#issuecomment-1454860316>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC5ITUY5M4PADNALNBJGHFDW2OOJBANCNFSM6AAAAAAVNLRZDA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

trzy commented:
Thanks a lot!

j-lag commented:
👍

Le jeu. 9 mars 2023, 19:35, trzy ***@***.***> a écrit :

> Thanks a lot!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/trzy/Supermodel/pull/64#issuecomment-1462575346>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC5ITUY35IX7766LXYIE7S3W3IPFZANCNFSM6AAAAAAVNLRZDA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
